modules/core/src/test/java/org/apache/ignite/cdc/ChangeDataCaptureSelfTest.java

```diff
@@ -0,0 +1,416 @@
+        IgniteCache<Integer, User> txCache = ign.getOrCreateCache(TX_CACHE_NAME);
+
+        addAndWaitForConsumption(
+            cnsmr,
+            cdc,
+            cache,
+            txCache,
+            ChangeDataCaptureSelfTest::addData,
+            0,
+            KEYS_CNT * 2,
+            getTestTimeout()
+        );
+
+        removeData(cache, 0, KEYS_CNT);
+
+        IgniteInternalFuture<?> rmvFut = runAsync(cdc);
+
+        assertTrue(waitForSize(KEYS_CNT, DEFAULT_CACHE_NAME, DELETE, getTestTimeout(), cnsmr));
+
+        rmvFut.cancel();
+
+        assertTrue(cnsmr.stopped());
+
+        stopAllGrids();
+
+        cleanPersistenceDir();
+    }
+
+    /** */
+    @Test
+    public void testReadBeforeGracefulShutdown() throws Exception {
+        IgniteConfiguration cfg = getConfiguration("ignite-0");
+
+        Ignite ign = startGrid(cfg);
+
+        ign.cluster().state(ACTIVE);
+
+        CountDownLatch cnsmrStarted = new CountDownLatch(1);
+        CountDownLatch startProcEvts = new CountDownLatch(1);
+
+        UserCDCConsumer cnsmr = new UserCDCConsumer() {
+            @Override public boolean onEvents(Iterator<ChangeDataCaptureEvent> evts) {
+                cnsmrStarted.countDown();
+
+                try {
+                    startProcEvts.await(getTestTimeout(), TimeUnit.MILLISECONDS);
+                }
+                catch (InterruptedException e) {
+                    throw new RuntimeException(e);
+                }
+
+                return super.onEvents(evts);
+            }
+        };
+
+        ChangeDataCapture cdc = new ChangeDataCapture(cfg, null, cdcConfig(cnsmr));
+
+        runAsync(cdc);
+
+        IgniteCache<Integer, User> cache = ign.getOrCreateCache(DEFAULT_CACHE_NAME);
+
+        addData(cache, 0, KEYS_CNT);
+
+        // Make sure all streamed data will become available for consumption.
+        Thread.sleep(2 * WAL_ARCHIVE_TIMEOUT);
+
+        cnsmrStarted.await(getTestTimeout(), TimeUnit.MILLISECONDS);
+
+        // Initiate graceful shutdown.
+        cdc.stop();
+
+        startProcEvts.countDown();
+
+        assertTrue(waitForSize(KEYS_CNT, DEFAULT_CACHE_NAME, UPDATE, getTestTimeout(), cnsmr));
+        assertTrue(waitForCondition(cnsmr::stopped, getTestTimeout()));
+
+        List<Integer> keys = cnsmr.data(UPDATE, cacheId(DEFAULT_CACHE_NAME));
+
+        assertEquals(KEYS_CNT, keys.size());
+
+        for (int i = 0; i < KEYS_CNT; i++)
+            assertTrue(keys.contains(i));
+    }
+
+    /** */
+    @Test
+    public void testMultiNodeConsumption() throws Exception {
+        IgniteEx ign1 = startGrid(0);
+
+        if (specificConsistentId)
+            consistentId = UUID.randomUUID();
+
+        IgniteEx ign2 = startGrid(1);
+
+        ign1.cluster().state(ACTIVE);
+
+        IgniteCache<Integer, User> cache = ign1.getOrCreateCache(DEFAULT_CACHE_NAME);
+
+        IgniteInternalFuture<?> addDataFut = runAsync(() -> addData(cache, 0, KEYS_CNT));
```

Comment:
If this check is about "CDC guarantees data processing even if data processed before the CDC start", a hint should be added here to prevent refactoring later.

Reply:
Right here, data added concurrently with CDC start.